% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/exported_functions.R
\name{mutate_msd}
\alias{mutate_msd}
\title{Calculate Primers for Multiple Site Saturation Mutagenesis}
\usage{
mutate_msd(
  input_sequence,
  codon = "NDT",
  prefix = "TT",
  restriction_enzyme = "GGTCTC",
  suffix = "A",
  vector = c("AATG", "AAGC"),
  replacements,
  replacement_range = 3,
  binding_min_length = 4,
  binding_max_length = 9,
  target_temp = 60,
  fragment_min_size = 100
)
}
\arguments{
\item{input_sequence}{The sequence which should be modified. This is an object of type character containing the sequence.}

\item{codon}{The desired type of MSD mutation [default: NDT]}

\item{prefix}{Additional nucleobases in 5' position of the recognition site [default: TT]}

\item{restriction_enzyme}{Recognition site sequence of the respective restriction enzyme [default: GGTCTC]}

\item{suffix}{Spacer nucleotides matching the cleavage pattern of the enzyme [default: A]}

\item{vector}{Four basepair overhangs complementary to the created overhangs in the acceptor vector  [default: c("AATG", "AAGC")]}

\item{replacements}{The desired substitutions. Can be a numeric vector or a list of character vectors with position and codon. See \href{articles/MSD3.html}{\code{vignette("MSD3", package = "goldenmutagenesis")}} for examples.}

\item{replacement_range}{Maximum distance in amino acid residues between two randomization sites to be incoporated into a single primer (reverse, end of the fragment) - has a cascading effect for following mutations [default: 3]}

\item{binding_min_length}{The minimal threshold value of the length of the template binding sequence in amino acid residues [default: 4]}

\item{binding_max_length}{Maximal length of the binding sequence in amino acid residues [default: 9]}

\item{target_temp}{Melting temperature of the binding sequence in \code{print('\u00B0')}C [default: 60]}

\item{fragment_min_size}{Minimal size of a generated gene fragment in base pairs [default 100]}
}
\value{
An object of class Primerset with the designed Primers.
}
\description{
The mutate_msd function designs the necessary set of primers for the desired mutations.
Note that you can also select TGG in saturation mutagenesis to apply the 22c trick.
}
\examples{
#Load the setup of the MSD vignette and design the primers
data(MSD_BsaI_setup_lv2)
print(mutations)
print(recognition_site_bsai)
primers<-mutate_msd(input_sequence, prefix="TT" ,
restriction_enzyme=recognition_site_bsai, suffix="A", 
vector=c("AATG", "AAGC"), replacements=mutations, replacement_range=5,
binding_min_length=4 , binding_max_length=9, target_temp=60,
fragment_min_size=60 )
}
